### PR TITLE
Added CanCancelStartingConnectionAfterNegotiate (#10432)

### PR DESCRIPTION
**Pull request by BalintFarkas**
Summary of the changes
Added test CanCancelStartingConnectionAfterNegotiate

Addresses #10432

### CODE REVIEW

File: src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.ConnectionLifecycle.cs [hunk 1]
```diff
@@ -380,6 +381,52 @@ public async Task SSEWaitsForResponseToStart()
...
+                        {
+                            // Kick off StartAsync, but don't wait for it
+                            var cts = new CancellationTokenSource();
+                            var startTask = connection.StartAsync(cts.Token);
+
+                            // Wait for the connection to get to the "WaitToContinue" call above,
+                            // which means it has gotten to Negotiate
+                            await negotiateSyncPoint.WaitForSyncPoint();
+
+                            // Assert that StartAsync has not yet been cancelled
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.ConnectionLifecycle.cs]: We use the other spelling of canceled :)

File: src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.ConnectionLifecycle.cs [hunk 2]
```diff
@@ -380,6 +381,52 @@ public async Task SSEWaitsForResponseToStart()
...
+                        CreateConnection(testHttpHandler),
+                        async (connection) =>
+                        {
+                            // Kick off StartAsync, but don't wait for it
+                            var cts = new CancellationTokenSource();
+                            var startTask = connection.StartAsync(cts.Token);
+
+                            // Wait for the connection to get to the "WaitToContinue" call above,
+                            // which means it has gotten to Negotiate
+                            await negotiateSyncPoint.WaitForSyncPoint();
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.ConnectionLifecycle.cs]: We generally put `.OrTimeout();` on all `await` calls in tests to help detect test hangs.

**Review by BrennanConroy — COMMENTED** [review]
This is a good start!

**Comment by halter73** [comment]
@BalintFarkas Thanks! This is a great start. If I understand correctly, this unit test verifies that HttpConnection.StartAsync(CancellationToken) flows the CancellationToken through to the HttpMessageHandler.SendAsync(HttpRequestMessage, CancellationToken) call associated with negotiate.

If you stop right now, we'll be plenty happy and make sure you get the contributor credit. That said, I think the other things we want to verify are:

1. A unit test that verifies the CancellationToken flows through to ITransport.StartAsync(Uri, TransferFormat, CancellationToken)
2. A unit test to verify that the CancellationToken passed to HubConnection.StartAsync(CancellationToken) flows through to IConnectionFactory.StartAsync(TransferFormat, CancellationToken)
3. A unit test to verify the ConnectionToken passed to HttpConnectionFactory.ConnectAsync(transferFormat, CancellationToken) flows through to HttpConnection.StartAsync(CancellationToken).
4. [Stretch] A unit test to verify the ConnectionToken passed to TcpConnectionFactory.ConnectAsync(CancellationToken) flows through to TcpConnection.StartAsync(CancellationToken). It doesn't do this today.
5. [Extra stretch] Functional tests to verify cancelling the token passed to ITransport.StartAsync(Uri, TransferFormat, CancellationToken) actually cancels the connection attempt for each transport (WS, SSE, LP).

I'm pretty sure we already test that the CancellationToken passed to HubConnection.StartAsync(CancellationToken) will cancel the hub handshake in a couple places so I think we're good thare.

**Comment by BalintFarkas** [comment]
@BrennanConroy  Fixed both, thanks!

@halter73 I'll stop now. I'm headed for a 10-day silent retreat which would almost certainly make this miss the end of June deadline. So I pulled this PR out of draft status and leave the rest of the matter in your capable hands. :) 

Thank you for everyone's support on this - as an ex-Microsoft employee it makes me proud to see how supportive Microsoft is toward contributors.

**Review by mikaelm12 — APPROVED** [review]
I think this looks fine. Let's merge this as is to get you in the contributor list, then add the rest of the test cases before closing the linked issue.

**Review by BrennanConroy — APPROVED** [review]

**Comment by BrennanConroy** [comment]
Thanks @BalintFarkas !

**Comment by BalintFarkas** [comment]
Sweet, thanks everyone! :)

**Pull request merged**